libs/crypto/merkle-impl/src/main/kotlin/net/corda/crypto/merkle/impl/MerkleProofImpl.kt

```diff
@@ -179,37 +174,39 @@ class MerkleProofImpl(
                         // Make new node with
-                        //   - left being current element, index $item.first, hash $item.second
+                        //   - left being current element
                         //   - right being a consumed incoming hash from $hashes[$hashIndex]
-                        //
-                        // Also remember we used $hashIndex by bumping the counter
-                        foundHashCallback(hashes[hashIndex], treeDepth+1, item.first + 1, hashIndex)
-                        val newHash = digest.nodeHash(treeDepth, item.second, hashes[hashIndex++])
-                        foundHashCallback( newHash, treeDepth, newIndex, null)
-                        Pair(newIndex, newHash)
+                        val newHash = digest.nodeHash(treeDepth, item.hash, hashes[hashIndex])
+                        MerkleNode(newIndex, newHash)
                     } else {
+                        // Remember we consumed a proof hash for the right hand child
+                        onNewHash(MerkleNodeInfo(MerkleNode(item.indexWithinLevel-1, hashes[hashIndex]), treeDepth+1, hashIndex))
                         // Make new node with:
                         //   - left being proof of hash at $hashIndex
                         //   - right being current element, index $item.first, hash $item.second
-                        //
-                        // Also remember we used hashIndex by bumping the counter.
-                        foundHashCallback(hashes[hashIndex], treeDepth+1, item.first-1, hashIndex)
-                        val newHash = digest.nodeHash(treeDepth, hashes[hashIndex++], item.second)
-                        foundHashCallback( newHash, treeDepth, newIndex, null)
-                        Pair(newIndex, newHash)
-                    }
+                        val newHash = digest.nodeHash(treeDepth, hashes[hashIndex], item.hash)
+                        MerkleNode(newIndex, newHash)
+                    })
+                    hashIndex++ // Remember we used an incoming hash
+                    onNewHash(MerkleNodeInfo(newNode, treeDepth, null))
+                    newItems.add(newNode)
                 } else {
-                    val newIndex = (item.first + 1) / 2
-                    foundHashCallback( item.second, treeDepth, newIndex, null)
                     // The last odd element, just gets lifted.
-                    newItems += Pair(newIndex, item.second)
+                    val newIndex = (item.indexWithinLevel + 1) / 2
+                    val newNode = MerkleNode(newIndex, item.hash)
+                    onNewHash(MerkleNodeInfo(newNode, treeDepth, null))
+                    newItems += MerkleNode(newIndex, item.hash)
                 }
-                ++index // whatever of the last 3 cases we took, we consumed one element
+                index++ // whatever of the last 3 cases we took, we consumed one element
```

Comment:
I wonder if this would be more understandable or not:
```suggestion
                index++ // whatever of the last 3 cases we took, we consumed one node
```
Just to distinguish it from the input hashes.
I am not too sure tbh.

Reply:
done